### DISCUSSION TITLE
How to preserve pandas index after query?

### DISCUSSION THREAD

**Discussion by fanurs** [duckdb/duckdb#3341] [General]
Hi authors,

Is it possible to preserve the indices after query, like `pandas.query()`? See below:
```python
import pandas as pd
import duckdb
df = pd.DataFrame({'n': [1, 2, 2], 'm': [0, 0, 1], 'x': [2, 3, 5], 'y': [5, 3, 2]})
df = df.set_index(['n', 'm'], drop=True)
print(duckdb.query('SELECT * FROM df WHERE x != 3').df()
print(df.query('x != 3'))
# Outputs:
# -------------
#    x  y
# 0  2  5
# 1  5  2
#      x  y
# n m      
# 1 0  2  5
# 2 1  5  2
```

My current "hack" is simply _not_ to drop the index columns, perform query using duckdb, then set the indices back. But I wonder if there is already a way to do it within duckdb APIs. Otherwise, could I inspire the team to add this tiny feature? 😄

Regards,
Fanurs

**Comment by Alex-Monahan** [comment]
Hello Fanurs!

The first challenge here is that there are many cases where the index will not be able to be maintained (for example, in a group by query that reduces the number of rows returned). As a result, I'm not sure if this would be a useful feature in very many cases.

If the index has valuable information, I believe you are already doing the workaround I would suggest! I would just convert the index to a column in your dataframe, query in DuckDB (including that column in your SELECT clause), return the results as a df, and then reset the index to that column in Pandas.

**Reply by hannes** [reply]
We could also consider dropping the index automatically for a copy of the data frame.
